geoprojbase: refactor everything to use the Projection abstraction

otan opened the pull request:
Previously, we use the projections map directly, which violates
abstraction bounds as we probably want to keep that map private. We
remedy this by moving everything behind the a Projection function
abstraction which prevents this leakage.

Furthermore, make Projection return an error directly. This is marked,
so we are able to reword error messages in other places.

Makes progress on #63969

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/63970)
<!-- Reviewable:end -->

sumeerbhola reviewed (changes requested):
> Reviewed 18 of 18 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63970)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan)

___
*[pkg/geo/geo.go, line 120 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwA9Zs2t8tlncHyJ0P:-MYwA9Zs2t8tlncHyJ0Q:bftqznl) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L120)):*
> ```Go
> 		if _, err := geoprojbase.Projection(spatialObject.SRID); err != nil {
> 			if errors.Is(err, geoprojbase.ErrProjectionNotFound) {
> 				return Geometry{}, errors.Newf("unknown SRID for Geometry: %d", spatialObject.SRID)
> ```

is there a reason we are changing the error string here -- isn't "projection for SRID %d does not exist" satisfactory?
___
*[pkg/geo/geo.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwAvh08nFlXWZcJZPz:-MYwAvh08nFlXWZcJZQ-:b-hkpbzf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L362)):*
> ```Go
> 		MaxY: math.MaxInt32,
> 	}
> 	if proj, err := geoprojbase.Projection(g.SRID()); err != nil {
> ```

how about strengthening the contract of geoprojbase.Projection to only return errors that are `ErrProjectionNotFound`? Or is the idea here that future implementations may encounter other errors?
___
*[pkg/geo/geo.go, line 365 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwB9uX-e9S-A3-Ihwi:-MYwB9uX-e9S-A3-Ihwj:b-6udftc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L365)):*
> ```Go
> 		if !errors.Is(err, geoprojbase.ErrProjectionNotFound) {
> 			return 0, err
> 		}
> ```

Why do we fall through for other errors instead of returning?
___
*[pkg/geo/geo.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwBLEACcihYv5-EgaB:-MYwBLEACcihYv5-EgaC:b-ioide0) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L391)):*
> ```Go
> 	lhs, err := g.SpaceCurveIndex()
> 	if err != nil {
> 		// The calling function from this does not have an error.
> ```

I didn't understand this statement?
Is this trying to say that the caller of Compare is expected to have ensured that `g.SpaceCurveIndex` will not return an error?

If our more complicated future implementations include keeping projection state on disk, we could have transient errors where the `ProjInfo` for an SRID is temporarily unavailable, so I am wary of panicing.
___
*[pkg/geo/geogen/geogen.go, line 334 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwCkuk4fRIqylzLHJd:-MYwCkuk4fRIqylzLHJe:b-6n8z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c103726de6994b26706244144d3d9f7dd6b1f58/pkg/geo/geogen/geogen.go#L334)):*
> ```Go
> 		randomBounds.minX, randomBounds.maxX = proj.Bounds.MinX, proj.Bounds.MaxX
> 		randomBounds.minY, randomBounds.maxY = proj.Bounds.MinY, proj.Bounds.MaxY
> 	}
> ```

This code seems to be handling the `!ok` case without panic, while the new code panics. Why the difference?
___
*[pkg/geo/geoprojbase/geoprojbase.go, line 86 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYw91KoE1KpOyIrTGAl:-MYw91KoE1KpOyIrTGAm:b8ey1v4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geoprojbase/geoprojbase.go#L86)):*
> ```Go
> var ErrProjectionNotFound error = errors.New("projection not found")
> 
> // Projection returns the ProjInfo identifier for the given SRID, as well as an
> ```

nit: ProjInfo identifier sounds odd, since ProjInfo is more than an identifier.


<!-- Sent from Reviewable.io -->

otan reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63970)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geo.go, line 362 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwAvh08nFlXWZcJZPz:-MYwT-2z-Cb_gj6EV6ea:bs1kktg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L362)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

how about strengthening the contract of geoprojbase.Projection to only return errors that are `ErrProjectionNotFound`? Or is the idea here that future implementations may encounter other errors?
</blockquote></details>

in future there may be more. in particular, if we read this from a file instead, we could have a file IO error.
___
*[pkg/geo/geo.go, line 365 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwB9uX-e9S-A3-Ihwi:-MYwT8dD-0srR4eLkApU:b-wpa747) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L365)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

Why do we fall through for other errors instead of returning?
</blockquote></details>

i've reworked this to make a bit more sense.
___
*[pkg/geo/geo.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwBLEACcihYv5-EgaB:-MYwTguTBglle5yn90ZL:bv88ghv) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L391)):*
> If our more complicated future implementations include keeping projection state on disk, we could have transient errors where the ProjInfo for an SRID is temporarily unavailable, so I am wary of panicing.

We either panic here or in tree/datum.go, which is both bad. I expect geometry to always be valid at this phase (we should read once from some buffer and store in memory forever, I was envisioning a lazy load), hence the panic.
___
*[pkg/geo/geogen/geogen.go, line 334 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwCkuk4fRIqylzLHJd:-MYwUL7p-JK2-_Nqi2KT:b-7sdc6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/1c103726de6994b26706244144d3d9f7dd6b1f58/pkg/geo/geogen/geogen.go#L334)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

This code seems to be handling the `!ok` case without panic, while the new code panics. Why the difference?
</blockquote></details>

we never generate a random geometry with an unknown SRID.
i am happy for this to go either way, chose panic for now. i don't think it makes too big a difference as it's only used for testing.


<!-- Sent from Reviewable.io -->

otan committed:
geoprojbase: refactor everything to use the Projection abstraction

Previously, we use the projections map directly, which violates
abstraction bounds as we probably want to keep that map private. We
remedy this by moving everything behind the a Projection function
abstraction which prevents this leakage.

Furthermore, make Projection return an error directly. This is marked,
so we are able to reword error messages in other places.

Release note: None

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 3 of 3 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-:-MYz168Q4QNFh_Wsi7BM:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @otan and @sumeerbhola)

___
*[pkg/geo/geo.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwBLEACcihYv5-EgaB:-MYz0_Y85Tpij8-Xt_qy:b-xujbyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L391)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

> If our more complicated future implementations include keeping projection state on disk, we could have transient errors where the ProjInfo for an SRID is temporarily unavailable, so I am wary of panicing.

We either panic here or in tree/datum.go, which is both bad. I expect geometry to always be valid at this phase (we should read once from some buffer and store in memory forever, I was envisioning a lazy load), hence the panic.
</blockquote></details>

I think what you are saying is that we may lazy load the `ProjInfo` but once loaded we will not be evicting it (not a cache), so if this `Geometry` has been successfully validated, which involved calling `Projection(<srid>)`, the subsequent call to the `Projection` function in `SpaceCurveIndex` should also succeed. Correct?


<!-- Sent from Reviewable.io -->

otan commented:
___
*[pkg/geo/geo.go, line 391 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/63970#-MYwBLEACcihYv5-EgaB:-MZEvkT51AVxq7Np1j75:b-9ir2qf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4a73a4f8bee97ef489108b43d4e6d9b27bd1b8a0/pkg/geo/geo.go#L391)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I think what you are saying is that we may lazy load the `ProjInfo` but once loaded we will not be evicting it (not a cache), so if this `Geometry` has been successfully validated, which involved calling `Projection(<srid>)`, the subsequent call to the `Projection` function in `SpaceCurveIndex` should also succeed. Correct?
</blockquote></details>

yep! that's the dream...


<!-- Sent from Reviewable.io -->

otan commented:
bors r=sumeerbhola
